bundles/org.eclipse.jface/src/org/eclipse/jface/viewers/ColumnViewer.java

```diff
@@ -817,4 +840,208 @@ protected void setBusy(boolean busy) {
 	public boolean isBusy() {
 		return busy;
 	}
+
+	@Override
+	protected Object[] getSortedChildren(Object parent) {
+		Object[] sorted = super.getSortedChildren(parent);
+		return applyItemsLimit(parent, sorted);
+	}
+
+	/**
+	 * Apply items limit for the items to be created on the viewer. This method is
+	 * supposed to be called at the end of {@link #getSortedChildren(Object)} call,
+	 * which always returns the complete list of elements to be populated.
+	 * <p>
+	 * <ul>
+	 * <li>If the {@link #setItemsLimit(int)} is not used by viewer, this method
+	 * does nothing.</li>
+	 * <li>If the {@link #setItemsLimit(int)} is used by viewer, this method might
+	 * modify the list by reducing number of elements according to the limit set,
+	 * and by adding an {@link IExpandableNode} as last element.</li>
+	 * <li>If the {@link #getSortedChildren(Object)} is not overridden by the
+	 * subclass, this method shouldn't be used by clients.</li>
+	 * </ul>
+	 * Note that in case of parent is {@link IExpandableNode} we will return next
+	 * block of limited elements to be created.
+	 *
+	 * @param parent parent element
+	 * @param sorted all children of given parent, as returned by
+	 *               {@link #getSortedChildren(Object)}
+	 * @return returns only limited items.
+	 * @since 3.31
+	 * @see ColumnViewer#setItemsLimit(int)
+	 */
+	protected final Object[] applyItemsLimit(Object parent, Object[] sorted) {
```

Comment:
Would it be possible to keep is package protected (thus not API)?

Reply:
fixed.